cpp/basix/cell.cpp

```diff
@@ -247,3 +248,182 @@ const std::string& cell::type_to_str(cell::type type)
   return it->second;
 }
 //-----------------------------------------------------------------------------
+double cell::volume(cell::type cell_type)
+{
+  switch (cell_type)
+  {
+  case cell::type::interval:
+    return 1;
+  case cell::type::triangle:
+    return 0.5;
+  case cell::type::quadrilateral:
+    return 1;
+  case cell::type::tetrahedron:
+    return 1.0 / 6;
+  case cell::type::hexahedron:
+    return 1;
+  case cell::type::prism:
+    return 0.5;
+  case cell::type::pyramid:
+    return 1.0 / 3;
+  default:
+    throw std::runtime_error("Unsupported cell type");
+  }
+  return 0;
+}
+//-----------------------------------------------------------------------------
+xt::xtensor<double, 2> cell::facet_outward_normals(cell::type cell_type)
+{
+  xt::xtensor<double, 2> normals = cell::facet_normals(cell_type);
+  xt::xtensor<bool, 1> orientations = cell::facet_orientations(cell_type);
+
+  for (std::size_t facet = 0; facet < normals.shape(0); ++facet)
+    if (orientations(facet))
+      for (std::size_t i = 0; i < normals.shape(1); ++i)
+        normals(facet, i) *= -1;
+
+  return normals;
+}
+//-----------------------------------------------------------------------------
+xt::xtensor<double, 2> cell::facet_normals(cell::type cell_type)
+{
+  const int tdim = cell::topological_dimension(cell_type);
+  xt::xtensor<double, 2> geometry = cell::geometry(cell_type);
+  std::vector<std::vector<int>> facets = cell::topology(cell_type)[tdim - 1];
+
+  xt::xtensor<double, 2> normals({facets.size(), (std::size_t)tdim});
+  for (std::size_t facet = 0; facet < facets.size(); ++facet)
```

Comment:
It seems a bit strange to me having the if statements inside the loop. It would be more natural to use a switch, with loops for each switch.